lib/add-channel.js

```diff
@@ -1,52 +1,85 @@
-const debug = require('debug')('semantic-release:github');
-const {RELEASE_NAME} = require('./definitions/constants');
-const parseGithubUrl = require('./parse-github-url');
-const resolveConfig = require('./resolve-config');
-const getClient = require('./get-client');
-const isPrerelease = require('./is-prerelease');
-
-module.exports = async (pluginConfig, context) => {
+import debugFactory from "debug";
+
+import { RELEASE_NAME } from "./definitions/constants.js";
+import parseGithubUrl from "./parse-github-url.js";
+import resolveConfig from "./resolve-config.js";
+import isPrerelease from "./is-prerelease.js";
+import { toOctokitOptions, SemanticReleaseOctokit } from "./octokit.js";
+
+const debug = debugFactory("semantic-release:github");
+
+export default async function addChannel(
+  pluginConfig,
+  context,
+  { Octokit = SemanticReleaseOctokit } = {}
+) {
   const {
-    options: {repositoryUrl},
+    options: { repositoryUrl },
     branch,
-    nextRelease: {name, gitTag, notes},
+    nextRelease: { name, gitTag, notes },
     logger,
   } = context;
-  const {githubToken, githubUrl, githubApiPathPrefix, proxy} = resolveConfig(pluginConfig, context);
-  const {owner, repo} = parseGithubUrl(repositoryUrl);
-  const octokit = getClient({githubToken, githubUrl, githubApiPathPrefix, proxy});
+  const { githubToken, githubUrl, githubApiPathPrefix, proxy } = resolveConfig(
+    pluginConfig,
+    context
+  );
+  const { owner, repo } = parseGithubUrl(repositoryUrl);
+  const octokit = new Octokit(
```

Comment:
to build on the injection question above, i'm curious if there would be value in passing an instance through rather than passing the constructor and constructing here and various other places. 

two potential benefits:
* would there be runtime benefits to using fewer instances throughout? it looks like the previous get-client approach still constructed a new instance each time, so this approach is pretty similar in this regard.
* would there be value in reducing the places where the instantiation logic exists by injecting an instance?

these are just questions for consideration. i'm good with either approach

Reply:
I had the same thought but didn't want to overburden this pull request. It would be an internal refactoring so shouldn't impact future releases.

Main benefit of having a shared `octokit` instance is that it would share the same throttling state. But it would be a side effect, which we already do for the `verified` flag here: [Line 12 in `/index.js`](https://github.com/semantic-release/github/blob/24dec8f6cc301648296528a861547cc564d345e5/index.js#L12)

We could add a singleton `octokit` instance. That is initiated first time any of the plugin methods are called? Only problem I see is that it would break if the different methods would be called with different configuration, but that doesn't seem to be a problem?

Do you have an idea how we might avoid the side effect for both `octokit` and `verified`? It's bound to cause trouble eventually, it aways does 🤣

Reply:
> I had the same thought but didn't want to overburden this pull request. It would be an internal refactoring so shouldn't impact future releases.

i think this is where my head is too. even if we do think it is a worthwhile change, it feels worthwhile to defer until after merging the rest of this change and releasing to latest.

would it be worth capturing an issue for later (optional) consideration? that way we are less likely to lose track of the thoughts you've shared here if we do decide to revisit?

Reply:
The only way I can think of is for semantic-release plugins to have a function that needs to be called, which then in turn returns the lifecycle functions. I guess we could add that functionality to `semantic-release` and support both patterns for a while or even forever. But for the time being we do the side effect?